src/main/cpp/src/datetime_parser.cu

```diff
@@ -0,0 +1,672 @@
+   * The bool in the returned tuple is false if the parse failed.
+   */
+  __device__ inline ParseResult parse_string_to_timestamp_us(
+    timestamp_components* ts_comp,
+    char const** parsed_tz_ptr,
+    size_type* parsed_tz_length,
+    cudf::string_view const& timestamp_str) const
+  {
+    const char* curr_ptr = timestamp_str.data();
+    const char* end_ptr  = curr_ptr + timestamp_str.size_bytes();
+
+    // trim left
+    while (curr_ptr < end_ptr && is_whitespace(*curr_ptr)) {
+      ++curr_ptr;
+    }
+    // trim right
+    while (curr_ptr < end_ptr - 1 && is_whitespace(*(end_ptr - 1))) {
+      --end_ptr;
+    }
+
+    if (curr_ptr == end_ptr) { return ParseResult::INVALID; }
+
+    const char* const bytes      = curr_ptr;
+    const size_type bytes_length = end_ptr - curr_ptr;
+
+    // segments stores: [year, month, day, hour, minute, seconds, microseconds, no_use_item,
+    // no_use_item] the two tail items are no use, but here keeps them as Spark does
+    int segments[]             = {1, 1, 1, 0, 0, 0, 0, 0, 0};
+    int segments_len           = 9;
+    int i                      = 0;
+    int current_segment_value  = 0;
+    int current_segment_digits = 0;
+    size_t j                   = 0;
+    int digits_milli           = 0;
+    // bool just_time = false;
+    thrust::optional<int> year_sign;
+    if ('-' == bytes[j] || '+' == bytes[j]) {
+      if ('-' == bytes[j]) {
+        year_sign = -1;
+      } else {
+        year_sign = 1;
+      }
+      j += 1;
+    }
+
+    while (j < bytes_length) {
+      char b           = bytes[j];
+      int parsed_value = static_cast<int32_t>(b - '0');
+      if (parsed_value < 0 || parsed_value > 9) {
+        if (0 == j && 'T' == b) {
+          // just_time = true;
+          i += 3;
+        } else if (i < 2) {
+          if (b == '-') {
+            if (!is_valid_digits(i, current_segment_digits)) { return ParseResult::INVALID; }
+            segments[i]            = current_segment_value;
+            current_segment_value  = 0;
+            current_segment_digits = 0;
+            i += 1;
+          } else if (0 == i && ':' == b && !year_sign.has_value()) {
+            // just_time = true;
+            if (!is_valid_digits(3, current_segment_digits)) { return ParseResult::INVALID; }
+            segments[3]            = current_segment_value;
+            current_segment_value  = 0;
+            current_segment_digits = 0;
+            i                      = 4;
+          } else {
+            return ParseResult::INVALID;
+          }
+        } else if (2 == i) {
+          if (' ' == b || 'T' == b) {
+            if (!is_valid_digits(i, current_segment_digits)) { return ParseResult::INVALID; }
+            segments[i]            = current_segment_value;
+            current_segment_value  = 0;
+            current_segment_digits = 0;
+            i += 1;
+          } else {
+            return ParseResult::INVALID;
+          }
+        } else if (3 == i || 4 == i) {
```

Comment:
These checks with constants against `i`, `b`, and `j` are pretty magical. Maybe some comments about what is going on and why? `// third and fourth characters should be digits from minutes` or some such.

Reply:
The parser code is ported from Spark, although has magic numbers, it's equivalent to Spark code.
Let's keep it them?